### PR TITLE
[IA-3106] Github action to Create Vm Image

### PR DESCRIPTION
**Pull request by jdcanas**
Github action to Create GCE and Dataproc VM images via a daisy workflow

successful build here: https://github.com/DataBiosphere/leonardo/actions/runs/3943212134/jobs/6747789691 

(not triggered on most recent commits due to the goal for it be triggered manually)

### CODE REVIEW

File: .github/workflows/custom_image_generation.yml [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+#             - 'jenkins/gce-custom-images/**' 
+
+#     - name: display progress
+#       run: | 
+#         echo "finished inspecting the repo to determine which image generation scripts to run@@@@@@@@@@@"
+
+jobs:
+  run-image-generation-script-gce:
+    runs-on: ubuntu-latest
+    # needs: inspect-changes
```
Comment (mikeflinn) [inline .github/workflows/custom_image_generation.yml]:
```suggestion
    permissions:
      contents: 'read'
      id-token: 'write'
    # needs: inspect-changes
```

You need to add this block in order for workload id federation to work

File: .github/workflows/custom_image_generation.yml [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        
+        # Make the image public
+        gcloud beta compute images add-iam-policy-binding \
+        projects/${GOOGLE_PROJECT}/global/images/${OUTPUT_BASE_NAME}-gce-cos-image-${GITHUB_SHA:0:6} \
+        --member='allAuthenticatedUsers' \
+        --role='roles/compute.imageUser'
+
+  run-image-generate-script-dataproc:
+    runs-on: ubuntu-latest
+    # needs: inspect-changes
```
Comment (mikeflinn) [inline .github/workflows/custom_image_generation.yml]:
```suggestion
    permissions:
      contents: 'read'
      id-token: 'write'
    # needs: inspect-changes
```
Same thing here

File: .github/workflows/custom_image_generation.yml [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+      - name: test step 2
+        run: echo "test 2"
+
+      - name: Create svc acct file
+        run: |
+          echo "$GCP_DAISY_GCR_KEY" > "$GITHUB_WORKSPACE/jenkins/dataproc-custom-images/$GOOGLE_CREDENTIAL_FILE_NAME"
+          echo 'echoing test name'
+          echo "$TEST_NAME"
+          echo 'catting'
+          cat $GITHUB_WORKSPACE/jenkins/dataproc-custom-images/$GOOGLE_CREDENTIAL_FILE_NAME
```
Comment (mikeflinn) [inline .github/workflows/custom_image_generation.yml]: Please never cat the contents of a service account file. Github actions logs are public, so any one who can see the leo repo could see this key. This should be done the same way as on line 74.

File: .github/workflows/custom_image_generation.yml [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+      - name: Init submodule
+        run: |
+          git submodule sync
+          git submodule update --init --recursive
+
+      - name: Run image generation script
+        run: |
+          docker run --rm -v $GITHUB_WORKSPACE/jenkins/dataproc-custom-images:/work:ro gcr.io/google.com/cloudsdktool/cloud-sdk:266.0.0 bash -c "cd /work/dataproc-custom-images \
+          && gcloud auth activate-service-account --key-file=/work/$GOOGLE_CREDENTIAL_FILE_NAME \
```
Comment (mikeflinn) [inline .github/workflows/custom_image_generation.yml]: I think this needs to be the same file path you setup in line 126. I don't think there is actually anything at this `/work/$GOOGLE_CREDENTIAL_FILE_NAME` path. Maybe a copy paste from Jenkins?

File: .github/workflows/custom_image_generation.yml (hunk 3 above)
Comment (jdcanas) [reply .github/workflows/custom_image_generation.yml]: ahh... this was just a sanity check, its not working currently, I just wanted to check if the file was indeed empty

File: .github/workflows/custom_image_generation.yml (hunk 4 above)
Comment (jdcanas) [reply .github/workflows/custom_image_generation.yml]: I'm fairly sure line 139 is mounting the directory with the key into the docker image's `work` dir. I'll remove the `:ro` though, not sure what that is.

Comment (mikeflinn) [reply .github/workflows/custom_image_generation.yml]: Ah yeah I misread it, my bad. `:ro` means read-only volume

File: .github/workflows/custom_image_generation.yml [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+      DATAPROC_IMAGE_VERSION_FORMATTED: 2-0-51-debian10
+      DATAPROC_IMAGE_BUCKET: gs://leo-dataproc-image-creation-logs
+      TEST_NAME: ${OUTPUT_BASE_NAME}-gce-cos-image-${GITHUB_SHA:0:6}
+      GCP_DAISY_GCR_KEY: ${{secrets.GCP_DAISY_GCR_KEY}}
+
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Create svc acct file
+        run: 'echo "$GCP_DAISY_GCR_KEY" > "$GITHUB_WORKSPACE/jenkins/dataproc-custom-images/$GOOGLE_CREDENTIAL_FILE_NAME"'
```
Comment (mikeflinn) [inline .github/workflows/custom_image_generation.yml]:
```suggestion
        run: 'echo "$GCP_DAISY_GCR_KEY" | base64 -d > "$GITHUB_WORKSPACE/jenkins/dataproc-custom-images/$GOOGLE_CREDENTIAL_FILE_NAME"'
```

**Comment by Qi77Qi** [comment]
where can we find the generated image name?

also can we delete [test.tmp](https://github.com/DataBiosphere/leonardo/pull/3088/files#diff-1886643ed4b95f41725a642a0e1bc314b4b4c1658b74dfabc1b0227b05117c76)

**Comment by jdcanas** [comment]
> where can we find the generated image name?
> 
> also can we delete [test.tmp](https://github.com/DataBiosphere/leonardo/pull/3088/files#diff-1886643ed4b95f41725a642a0e1bc314b4b4c1658b74dfabc1b0227b05117c76)

Deleted, and I added a step OUTPUT_IMAGE_URI that prints it. Unfortunately, there's not a good way to display it on the github actions UI that I found with my cursory googling..

**Review by n8landolt — APPROVED** [review]

**Comment by jdcanas** [comment]
jenkins retest

**Review by Qi77Qi — APPROVED** [review]
looks great.. :shipit:

**Comment by jdcanas** [comment]
jenkins retest

**Pull request merged**